lib/backend/dbi.c

```diff
@@ -97,7 +98,7 @@ dbDetectBackend(rpmdb rdb)
 		rpmlog(RPMLOG_WARNING,
 			_("Converting database from %s to %s backend\n"),
 			ondisk->name, cfg->name);
-	    } else {
+	    } else if (*db_backend) {
```

Comment:
Does it make sense here to have no final `else` case? This seems not right...

Reply:
But it is intended. Panu may wish to have a rpmlog(RPMLOG_DEBUG,...) call in that case, though.

Reply:
Right, it does look a kinda suspect that way.

I can see how rpmdb2solv doesn't want that warning to be emitted, because it *wants* rpm to do a probe. But shouldn't this actually be:

```
/* On-disk database differs from configuration */
if (ondisk && ondisk != cfg) {
    if (cfg) {
        if (rdb->db_flags & RPMDB_FLAG_REBUILD) {
            rpmlog(RPMLOG_WARNING,
                    _("Converting database from %s to %s backend\n"),
                    ondisk->name, cfg->name);
        } else {
            rpmlog(RPMLOG_WARNING,
                _("Found %s %s database while attempting %s backend: "
                "using %s backend.\n"),
                ondisk->name, ondisk->path, db_backend, ondisk->name);
        }
    }
    rdb->db_ops = ondisk;
}   

```

Ie the warnings should only ever be emitted when there *is* a configuration, and I think conditionalizing on 'cfg' makes it more clear than the various different conditions (eg the `cfg->name` access *looks* like potential a NULL-pointer dereference now, that its guarded by RPMDB_FLAG_REBUILD is not immediately obvious, and looking at *db_backend is different from all the other surrounding conditions). Or am I missing something? (this is a surprisingly subtle little thing)?

Reply:
With your code there will no longer be a warning for unknown but configured backends in read-only mode. We can also catch this in the invalid _db_backend test above if you want this to error out instead. But if we allow it I think we should print a warning.

Reply:
We can use
```
if (ondisk && ondisk != cfg) {
    if (*db_backend) {
        if (rdb->db_flags & RPMDB_FLAG_REBUILD) {
        ...
```
and change the `cfg->name` to `db_backend` just for consistency. Would that be more to your taste?

Reply:
Conditionalizing and using db_backend instead sounds fine to me. Might also want to update the comments to explain what exactly is the intention here, because like said, it's surprisingly subtle.
That debug log note in the case it goes on auto-probe in the non-warning case wouldn't hurt either.